constraints/src/test/java/gov/nasa/jpl/aerie/constraints/time/WindowsTest.java

```diff
@@ -1,559 +1,506 @@
 
-    final var expected = new Windows();
 
-    assertEquivalent(expected, windows);
+    Windows expected = new Windows(Segment.of(interval(-1, 1, SECONDS), true),
+                                   Segment.of(interval(1, Exclusive, 2, Inclusive, SECONDS), false),
+                                   Segment.of(interval(4, Inclusive, 11, Exclusive, SECONDS), true),
+                                   Segment.of(interval(12, 15, SECONDS), true),
+                                   Segment.of(interval(15, Exclusive, 16, Exclusive, SECONDS), false),
+                                   Segment.of(interval(Duration.MAX_VALUE.minus(Duration.of(2, SECONDS)),
+                                                       Duration.MAX_VALUE), true));
+    assertIterableEquals(expected, result);
   }
 
   @Test
-  public void intersectAdjacent() {
-    final var windows = new Windows();
-    windows.add(window(0,  1, MICROSECONDS));
-    windows.add(window(2,  4, MICROSECONDS));
+  public void shiftByFromStartFromEndPermsWithInterval() {
+    var leftEnd = interval(0, 2, SECONDS);
+    var rightEnd = interval(8, 10, SECONDS);
 
-    windows.intersectWith(1,  2, MICROSECONDS);
+    var orig = new Windows()
+        .set(new Windows(Segment.of(leftEnd, true), Segment.of(rightEnd, true)));
 
-    final var expected = new Windows();
-    expected.addPoint(1, MICROSECONDS);
-    expected.addPoint(2, MICROSECONDS);
+    var fromStartPosFromEndPos = orig.shiftBy(Duration.of(-1, SECONDS), Duration.of(1, SECONDS));
+    assertIterableEquals(fromStartPosFromEndPos, new Windows(
+        Segment.of(interval(-1, 3, SECONDS), true),
+        Segment.of(interval(7, 11, SECONDS), true)
+    ));
 
-    assertEquivalent(expected, windows);
-  }
+    var fromStartPosFromEndNeg = orig.shiftBy(Duration.of(1, SECONDS), Duration.of(-1, SECONDS));
+    assertEquals(fromStartPosFromEndNeg, new Windows(
+        Segment.of(interval(0, Inclusive, 1, Exclusive, SECONDS), false),
+        Segment.of(interval(1, 1, SECONDS), true),
+        Segment.of(interval(1, Exclusive, 2, Inclusive, SECONDS), false),
+        Segment.of(interval(8, Inclusive, 9, Exclusive, SECONDS), false),
+        Segment.of(interval(9, 9, SECONDS), true),
+        Segment.of(interval(9, Exclusive, 10, Inclusive, SECONDS), false)
+    ));
 
-  @Test
-  public void intersectNonintersecting() {
-    final var windows = new Windows();
-    windows.add(window(0,  1, MICROSECONDS));
-    windows.add(window(2,  4, MICROSECONDS));
+    var fromStartNegFromEndPos = orig.shiftBy(Duration.of(1, SECONDS), Duration.of(1, SECONDS));
+    assertEquals(fromStartNegFromEndPos, new Windows(
+        Segment.of(interval(0, Inclusive, 1, Exclusive, SECONDS), false),
+        Segment.of(interval(1, 3, SECONDS), true),
+        Segment.of(interval(8, Inclusive, 9, Exclusive, SECONDS), false),
+        Segment.of(interval(9, 11, SECONDS), true)
+    ));
```

Comment:
This specification seems a little ill-founded relative to how we think about missing information. If we shift both start and end by some positive amount `d` forward, we're just translating all values forward. The value at time `t` becomes the value at time `t + d`. But this specification seems to say that the value at time 8 ought to be *false*, regardless of what the value at time 7 might have been if we were to resolve the null that was there previously.

Reply:
Yeah... I don't know how to make this form of `shiftBy` compatible with nulls. I originally argued for this generalized operator in a context without nulls; now maybe it makes more sense to have a separate `translate` operator (which is well-founded). The uses of `shiftBy` to expand/contract "windows" seem *very suspicious* to me, and much more in line with a Spans use-case...

Reply:
There is also a potential `convolve` operator that could be well-defined on windows, and would support use-cases for expanding true windows to the left and right. There's probably a dual for contracting true windows / expanding false windows. 🤔

Reply:
Ok. Sorry this took two weeks, I exist on Aerie again now.

While I agree that shiftBy is much more easily defined and intuitive on Spans, unfortunately it will be used very heavily in Constraints. Expressions like "during this condition and for five minutes after", or "for all but the first five minutes of this condition" should be perfectly accessible in a Windows context, because forcing the user to convert to Spans and back will lose any null information from the resource profiles they based their expression on.

The new implementation for shifting Windows will produce the same result as before, for Windows with no nulls. And I've reinterpreted the procedure to be valid in a context with nulls:

1. leave all `false` segments as they are
2. shift `true` segments according to the arguments
3. `true` segments leave `false` behind if there are uncovered segments that used to be `true`

The actual algorithm for this is as follows:

1. set the value of all defined segments to `false`.
2. overwrite segments with `true` according to the shifted windows.

I can't think of a better result, and unfortunately I don't think forcing conversion to Spans is an option. I don't like it either, but is this at least acceptable?

Reply:
Yeah... I'm not going to block the PR on this, I just feel really uncomfortable about how it doesn't match our expectations of `null`. It feels like a lurking footgun, since it's named very intuitively and yet arguably has semantics that would surprise a user.

The version of `shiftBy` that I wrote (which doesn't meet the tests) is this:

```java
  public Windows shiftBy(Duration fromStart, Duration fromEnd){
    final var builder = IntervalMap.<Boolean>builder();

    for (final var segment : this.segments) {
      final var interval = segment.interval();

      final var shiftedInterval = Interval.between(
            interval.start.plus(fromStart), interval.startInclusivity,
            interval.end.plus(fromEnd),     interval.endInclusivity);

      builder.add(shiftedInterval, segment.value());
    }

    return new Windows(builder.build());
  }
```

It's a bit like a restricted `convolve` operator (which I also have an implementation of [here](https://gist.github.com/Twisol/e82110fe3031d3f8bc2639adbb768f0d)). It applies uniformly to every segment (even false ones), and I think it respects our expectations for `null`. (In particular, we don't leave snail trails of `false`s if the preceding interval is `null`.)

> "during this condition and for five minutes after"

This could be implemented as `condition.convolve(0, 5, MINUTES)`, I think?

> "for all but the first five minutes of this condition"

If a `true` interval is preceded by a `null` interval, we don't actually know that that's when it was first `true`. In fact, two `true`s separated by a `null` might merge when those `null`s are resolved, meaning the entire second interval should have been excluded in the result.

Setting that aside, how does `shiftBy` satisfy this use-case? At first I thought `condition.minus(condition.translate(+5, MINUTES))` would work, but if any two intervals are small enough and within 5 minutes of each other, the advanced shadow of the earlier one might subtract out the later one, which would not be a correct solution for this use-case.

On the other hand, if we're only looking for the single *global* first time the condition is true, it could be implemented with `condition.minus(condition.convolve(0, INFINITY).translate(+5, MINUTES))`. The convolution basically detects the first transition, and sets everything into the future to `true`; the translation uncovers the first five minutes, and then we subtract everything later off.

Reply:
I've edited my comment above a grand total of 19 times, if that gives any indication of my thought process.

Reply:
My implementation above has the *very* nice property that `condition.not(a, b).negate() == condition.not().shiftBy(b, a)`, by the by. Which makes it super clear that your two use-cases are dual, since we're just extending either the true intervals or the false intervals by 5 minutes. 👀

Reply:
I'm not sure what you mean by `not(a,b)` in your last comment.

Reply:
Ah, typo. I assume you mean

> My implementation above has the very nice property that `condition.shiftBy(a, b).not() == condition.not().shiftBy(b, a)`

In which case that would be pretty nifty. I'll try it

Reply:
yes, whoops -- sorry. I wrote what you quoted but with `negate` instead of `not` originally, then noticed that the method was actually called `not`. Apparently I accidentally replaced one `shiftBy` with `not` instead of replacing the adjacent `negate`.

Reply:
Ah now that I think about it, this was the approach I originally tried to implement, and decided I didn't like it.

For starters, it doesn't actually commute with `not`. For example, if the Windows is alternating between `false` and gaps, with no `true` segments, then `shiftBy(...)` should do nothing. But `.not().shiftBy(...).not()` will do quite a lot.

Also, the behavior for "leaving" behind `false` or `null` depending on its neighbors can break, if you shrink a window until it disappears, it is neighbored by a `false` and `null` on each side.

And honestly, none of the possible algorithms work as intended when `null` meets `true`, because its not possible to behave correctly. Gap are supposed to mean "unknown", not "undefined", so in reality the true segment could have started or ended inside a gap; so the ends should be shifted by an unknown amount.

`shiftBy` seems totally incompatible with the new definition of Windows, but we need it to exist anyway. So I picked an algorithm that in effect blindly pretends like its using the old definition of Windows.

Reply:
> For example, if the Windows is alternating between `false` and gaps, with no `true` segments, then `shiftBy(...)` should do nothing. But `.not().shiftBy(...).not()` will do quite a lot.

I don't think I agree with this logic. Gaps are nulls; nulls represent missing information. There could absolutely be unknown true regions hiding in those gaps, so `false` intervals must shrink if `true` intervals expand. The logic is entirely dual for alternating `true` and `null`, where shrinking `true` windows means expanding potential `false` windows.

> Also, the behavior for "leaving" behind `false` or `null` depending on its neighbors can break, if you shrink a window until it disappears, it is neighbored by a `false` and `null` on each side.

Can you detail what you mean by "break" here? I'm not able to visualize an example based on what you've said here.

> so in reality the true segment could have started or ended inside a gap; so the ends should be shifted by an unknown amount.

The degree of uncertainty is captured by the nulls that remain after the shift. If `true` becomes `null` at time `t`, and we want to extend end times by a delta `d`, then we do certainly know that everything from `t` to `t + d` will now be true. The null interval shrinks; but anything from `t+d` onward in that interval *remains null*, because it could have been extended further depending on how the original `null` resolves.

I'm actually totally flipped on this now -- I think the definition of `shiftBy` I gave above is semantically well-founded, even in a `Windows` context. 🤔

Reply:
whoops, typo:

>  if you shrink a window until it disappears, *AND* it is neighbored by a false and null on each side.

If we have `[-inf, 0): false, [0, 1): true, [1, inf): gap` and we `.shiftBy(2, -2)`, what is the result?

Reply:
> For starters, it doesn't actually commute with `not`.

I'm going to try to write a property test to check this (with `jqwik`). I'm almost completely confident it does commute in the described way (with the arguments swapped)...

Reply:
> whoops, typo:
> 
> > if you shrink a window until it disappears, _AND_ it is neighbored by a false and null on each side.
> 
> If we have `[-inf, 0): false, [0, 1): true, [1, inf): gap` and we `.shiftBy(2, -2)`, what is the result?

`[-inf, 3]`. The second argument never applies, because the true interval never ends. Another way to think of the two arguments is "deltaFalseToTrue` and `deltaTrueToFalse`.

Reply:
Crud, I completely misread your message. Just ignore that one above for a moment.

Reply:
[-inf, 0): false
[0, 1): true
[1, inf): gap

[-inf, +inf]
[-inf, 2] , [2, inf]